compiler/lib/parse_js.ml

```diff
@@ -19,115 +19,165 @@
+            , lexbuf.Lexing.lex_curr_p )
+        in
+        loop_error prev checkpoint
+    | I.Shifting _ | I.AboutToReduce _ -> loop_error prev (I.resume checkpoint)
+    | I.Accepted _ -> assert false
+    | I.Rejected -> `Error prev
+    | I.HandlingError _ -> loop_error prev (I.resume checkpoint)
+  in
+  let rec loop prev comments (inputneeded, checkpoint) =
+    let module I = Js_parser.MenhirInterpreter in
+    match checkpoint with
+    | I.InputNeeded _env ->
+        let inputneeded = checkpoint in
+        let token, comments =
+          match prev with
+          | (Js_token.EOF _ as prev) :: _ -> prev, comments
+          | _ ->
+              let rec read_one comments lexbuf =
+                match Js_lexer.main lexbuf with
+                | (TCommentLineDirective _ | TComment _) as t ->
+                    read_one (t :: comments) lexbuf
+                | t -> t, comments
+              in
+              let t, comments = read_one comments lexbuf in
+              let t =
+                match prev, t with
+                (* restricted productions
+                 * 7.9.1 - 3
+                 * When, as the program is parsed from left to right, a token is encountered
+                 * that is allowed by some production of the grammar, but the production
+                 * is a restricted production and the token would be the first token for a
+                 * terminal or nonterminal immediately following the annotation [no LineTerminator here]
+                 * within the restricted production (and therefore such a token is called a restricted token),
+                 * and the restricted token is separated from the previous token by at least
+                 * one LineTerminator, then a semicolon is automatically inserted before the
+                 * restricted token. *)
+                | ( (T_RETURN _ | T_CONTINUE _ | T_BREAK _ | T_THROW _) :: _
+                  , ((T_SEMICOLON _ | T_VIRTUAL_SEMICOLON _) as t) ) ->
+                    t
+                | (T_RETURN _ | T_CONTINUE _ | T_BREAK _ | T_THROW _) :: _, t
+                  when fol prev t ->
+                    reset lexbuf;
+                    T_VIRTUAL_SEMICOLON Parse_info.zero
+                (* The practical effect of these restricted productions is as follows:
+                 * When a ++ or -- token is encountered where the parser would treat it
+                 * as a postfix operator, and at least one LineTerminator occurred between
+                 * the preceding token and the ++ or -- token, then a semicolon is automatically
+                 * inserted before the ++ or -- token. *)
+                | _, (T_DECR cpi as tok) when not (fol prev tok) -> Js_token.T_DECR_NB cpi
+                | _, (T_INCR cpi as tok) when not (fol prev tok) -> Js_token.T_INCR_NB cpi
+                | _, ((T_DIV _ | T_DIV_ASSIGN _) as tok) ->
+                    if I.acceptable checkpoint tok lexbuf.Lexing.lex_start_p
```

Comment:
What does this condition mean?

Reply:
This is part of the incremental api. It checks if a token is valid given a parser state (checkpoint).
Div and Regexp tokens and mutually exclusive in the JavaScript parser. We first check if div is an acceptable token, if not, we reset the lexer and lex a Regexp instead